pennylane/ops/op_math/exp.py

```diff
@@ -240,11 +266,115 @@ def decomposition(self):
         Returns:
             list[PauliRot]: decomposition of the operator
         """
-        if not self.has_decomposition:
-            raise DecompositionUndefinedError
-        new_coeff = math.real(2j * self.coeff)
-        string_base = qml.pauli.pauli_word_to_string(self.base)
-        return [qml.PauliRot(new_coeff, string_base, wires=self.wires)]
+        return self._recursive_decomposition(self.base, self.coeff)
+
+    def _recursive_decomposition(self, base: Operator, coeff: complex):
+        """Decompose the exponential of ``base`` multiplied by ``coeff``.
+
+        Args:
+            base (Operator): exponentiated operator
+            coeff (complex): coefficient multiplying the exponentiated operator
+
+        Returns:
+            List[Operator]: decomposition
+        """
+        if isinstance(base, Tensor) and len(base.wires) != len(base.obs):
+            raise DecompositionUndefinedError(
+                "Unable to determine if the exponential has a decomposition "
+                "when the base operator is a Tensor object with overlapping wires. "
+                f"Received base {base}."
+            )
+
+        with qml.QueuingManager.stop_recording():
+            # Change base to `Sum`/`Prod`
+            if isinstance(base, Hamiltonian):
+                base = qml.dot(base.coeffs, base.ops)
+            elif isinstance(base, Tensor):
+                base = qml.prod(*base.obs)
+
+        if isinstance(base, SProd):
+            return self._recursive_decomposition(base.base, base.scalar * coeff)
+
+        for op_class in self._ops_with_generator:
+            # Check if the exponentiated operator is a generator of another operator
+            if op_class.num_wires == base.num_wires:
+                with qml.QueuingManager.stop_recording():
+                    g, c = qml.generator(op_class)(coeff, base.wires)
+                    # Make sure both ``g`` and ``base`` are of the same type
+                    if isinstance(g, Hamiltonian):
+                        g = qml.dot(g.coeffs, g.ops)
+                    elif isinstance(g, Tensor):
+                        g = qml.prod(*g.obs)
```

Comment:
I wonder if we should add a new mode to `qml.generator` to absorb this logic into the `qml.generator` function itself.

Reply:
We definitely can, but this is outside the scope of this PR. I can create an issue if you want.

Reply:
Edit: sorry I misunderstood what you were asking for

Moved this inside `_generator_prefactor`, such that now it always returns a `Sum` and `Prod` instead of a `Hamiltonian` and `Tensor`.